metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -375,64 +341,223 @@ void convert_batch_to_row_diff(const std::string &graph_fname,
+                nullptr, std::move(matrix), graph_fname + anchors_extension);
+        row_diff[l_idx] = std::make_unique<RowDiffAnnotator>(std::move(diff_annotation),
+                                                             label_encoders[l_idx]);
         auto fname = std::filesystem::path(source_files[l_idx])
                 .filename()
                 .replace_extension()
                 .replace_extension(RowDiffAnnotator::kExtension);
         auto fpath = dest_dir/fname;
-        annotator.serialize(fpath);
+        row_diff[l_idx]->serialize(fpath);
         logger->trace("Serialized {}", fpath);
     }
     logger->trace("Removing temp directory: {}", tmp_path);
     std::filesystem::remove_all(tmp_path);
+
+    sdsl::int_vector_buffer row_reduction(row_reduction_fname, std::ios::in | std::ios::out, 1024 * 1024);
+    uint64_t num_larger_rows = 0;
+    ProgressBar progress_bar(row_reduction.size(), "Update row reduction",
+                             std::cerr, !common::get_verbose());
+    for (node_index chunk = 0; chunk < row_reduction.size(); chunk += BLOCK_SIZE) {
+        row_nbits_batch.assign(std::min(BLOCK_SIZE, row_reduction.size() - chunk), 0);
+
+        #pragma omp parallel for num_threads(num_threads) schedule(dynamic)
+        for (size_t l_idx = 0; l_idx < row_diff.size(); ++l_idx) {
+            const auto &columns = row_diff[l_idx]->get_matrix().diffs().data();
+            for (size_t j = 0; j < columns.size(); ++j) {
+                const std::unique_ptr<bit_vector> &source_col = columns[j];
+                source_col->call_ones_in_range(chunk, chunk + row_nbits_batch.size(),
+                    [&](node_index i) {
+                        __atomic_add_fetch(&row_nbits_batch[i - chunk], 1, __ATOMIC_RELAXED);
+                    }
+                );
+            }
+        }
+
+        __atomic_thread_fence(__ATOMIC_ACQUIRE);
+
+        for (uint64_t i = 0; i < row_nbits_batch.size(); ++i) {
+            row_reduction[chunk + i] -= row_nbits_batch[i];
+            // check if the row reduction is negative
+            if (row_reduction[chunk + i] >> (row_reduction.width() - 1))
+                num_larger_rows++;
+        }
+
+        progress_bar += row_nbits_batch.size();
+    }
+
+    logger->trace("Rows with negative row reduction: {} in vector {}",
+                  num_larger_rows, row_reduction_fname);
+}
+
+void optimize_anchors_in_row_diff(const std::string &graph_fname,
+                                  const std::filesystem::path &dest_dir,
+                                  const std::string &row_reduction_extension) {
+    if (std::filesystem::exists(graph_fname + ".terminal")) {
+        logger->info("Found optimized anchors {}", graph_fname + ".terminal");
+        return;
+    }
+
+    logger->trace("Optimizing anchors");
+
+    std::vector<std::string> filenames;
+    for (const auto &p : std::filesystem::directory_iterator(dest_dir)) {
```

Comment:
this will take quite a bit of time for datasets like bigsi

Reply:
yeah, hm, but not more than ls I guess, so not that much overall